tensorflow/stream_executor/cuda/cuda_dnn.cc

```diff
@@ -333,16 +333,60 @@ port::Status GetLoadedCudnnVersion(CudnnVersion* version) {
   return ::tensorflow::OkStatus();
 }
 
-#if CUDNN_MAJOR >= 8 && (CUDNN_MINOR > 0 || CUDNN_PATCHLEVEL >= 4)
-void PreloadCudnnLibrary(cudnnStatus_t (*version_check_fn)(),
-                         absl::string_view sub_library) {
-  cudnnStatus_t status = version_check_fn();
-  if (status != CUDNN_STATUS_SUCCESS) {
-    VLOG(1) << "Could not pre-initialize cuDNN sub-library " << sub_library
-            << ".  Error: " << cudnnGetErrorString(status) << ".";
+enum class PreloadCudnnType { ConvFwd, ConvBwdFilter, ConvBwdData, Rnn };
+
+// Preload sub libs for cudnn 8.0.4+
+void PreloadCudnnSubLibs(PreloadCudnnType type) {
+#if CUDNN_VERSION >= 8004
+  switch (type) {
+    case PreloadCudnnType::ConvBwdFilter: {
+      [[clang::fallthrough]];
+    }
+    case PreloadCudnnType::ConvBwdData: {
+      cudnnOpsTrainVersionCheck();
+      cudnnCnnTrainVersionCheck();
+      [[clang::fallthrough]];
+    }
+    case PreloadCudnnType::ConvFwd: {
+      cudnnOpsInferVersionCheck();
+      cudnnCnnInferVersionCheck();
+      break;
+    }
+    case PreloadCudnnType::Rnn: {
+      cudnnOpsInferVersionCheck();
+      cudnnAdvInferVersionCheck();
+      cudnnOpsTrainVersionCheck();
+      cudnnAdvTrainVersionCheck();
+      break;
+    }
+    default: {
+      LOG(WARNING) << "Unsupported cuDNN preload type.";
+      break;
```

Comment:
I would recommend removing this (default) in order to delegate this check to the compiler (the compiler will warn, or issue an error in case of `-wall`, if a case is not handled as mentioned [here](https://github.com/tensorflow/tensorflow/pull/56656#discussion_r912421814)).